### PR TITLE
feat: Promote modifications to public API

### LINKED ISSUE #1652: Promote `model._modifications` to public API
### Summary

[`pyhf.pdf._MainModel._modifications`](https://github.com/scikit-hep/pyhf/blob/be5ae6561d482d18601c9301f28f7b242b5a7a2f/src/pyhf/pdf.py#L529-L543) is useful for model visualization and debugging. See for example https://github.com/scikit-hep/pyhf/pull/1639#discussion_r730752320. It also came up in the discussion https://github.com/scikit-hep/pyhf/discussions/1414. This function is currently not part of the public API. Given its usefulness, I propose changing that. This suggestion also came up in https://github.com/scikit-hep/pyhf/pull/1461#discussion_r633482129.

Two alternatives are possible: using the `_modifications` API even though it signals it is not public (accepting a possible breaking change in the future), or re-implementing it externally.

### Additional Information

See https://github.com/scikit-hep/pyhf/pull/1639#discussion_r730752320 for example usage, this API could presumably also be used for [`pyhf-validation/scripts/validate_systs.py`](https://github.com/pyhf/pyhf-validation/blob/master/scripts/validate_systs.py).

### Code of Conduct

- [X] I agree to follow the Code of Conduct

### PR DESCRIPTION
**Pull request by kratsg**
# Pull Request Description

Resolves #1652.

Docs: https://pyhf--2339.org.readthedocs.build/en/2339/_generated/pyhf.pdf.Model.html#pyhf.pdf.Model

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR
```
* Promote `pyhf.pdf._MainModel._modifications` to `pyhf.pdf._MainModel.modifications`.
* Promote `pyhf.pdf.Model._modifications` to `pyhf.pdf.Model.modifications`.
* Add `pyhf.pdf._MainModel` and `pyhf.pdf._ConstraintModel` to the docs.
* Add documentation for the `modifications()` functions.
```

### CODE REVIEW

File: src/pyhf/pdf.py [hunk 1]
```diff
@@ -650,7 +650,21 @@ def logpdf(self, maindata, pars):
...
+
+        Args:
+            pars (:obj:`tensor`): The model parameters
+
+        Returns:
+            Tuple of additive and multiplicative modifications:
+                - deltas (:obj:`list`) is the result of an `apply(pars)` of combined modifiers
+                  with 'addition' op_code
+                - factors (:obj:`list`) is the result of `apply(pars)` of combined modifiers
+                  with 'multiplication' op_code
```
Comment (matthewfeickert) [inline src/pyhf/pdf.py:666]: I realize that you were just taking the style from the rest of this file, but this file is old and doesn't really work with our current rendering, so to make things appear code-like we can just throw on some additional backticks:

```suggestion
                - deltas (:obj:`list`) is the result of an ``apply(pars)`` of combined modifiers
                  with ``"addition"`` ``op_code``
                - factors (:obj:`list`) is the result of ``apply(pars)`` of combined modifiers
                  with ``"multiplication"`` ``op_code`
```

**Review by matthewfeickert — COMMENTED** [review]
Thanks @kratsg. I just have a docs styling comment, but other than that LGTM.

**Comment by matthewfeickert** [comment]
(Though maybe we also wait to see if @alexander-held has any comments before merge, given that he opened up Issue #1652)

**Review by alexander-held — COMMENTED** [review]
Nothing else from me, thanks for addressing this!

File: src/pyhf/pdf.py (hunk 1 above)
Comment (matthewfeickert) [reply src/pyhf/pdf.py:666]: Render of the suggested changes:

image

File: src/pyhf/pdf.py [hunk 2]
```diff
@@ -650,7 +650,21 @@ def logpdf(self, maindata, pars):
...
+
+        Args:
+            pars (:obj:`tensor`): The model parameters
+
+        Returns:
+            Tuple of additive and multiplicative modifications:
+                - deltas (:obj:`list`) is the result of an ``apply(pars)`` of combined modifiers
+                  with ``"addition"`` ``op_code``
+                - factors (:obj:`list`) is the result of ``apply(pars)`` of combined modifiers
+                  with ``"multiplication"`` ``op_code`
```
Comment (matthewfeickert) [inline src/pyhf/pdf.py:666]:
```suggestion
                  with ``"multiplication"`` ``op_code``
```

My bad I forgot a "`". Adding here and then commmiting.

**Review by matthewfeickert — APPROVED** [review]
LGTM. Thanks @kratsg!

**Pull request merged**